src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceSetDeserializer.cs

```diff
@@ -124,10 +126,133 @@ public virtual IEnumerable ReadResourceSet(ODataResourceSetWrapper resourceSet,
+
+            IEdmModel model = readContext.Model;
+
+            if (model == null)
+            {
+                throw Error.Argument("readContext", SRResources.ModelMissingFromReadContext);
+            }
+
+            IEdmStructuredType actualType = model.FindType(deltalink.TypeAnnotation.TypeName) as IEdmStructuredType;
+            if (actualType == null)
             {
-                yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                throw new ODataException(Error.Format(SRResources.ResourceTypeNotInModel, deltalink.TypeAnnotation.TypeName));
             }
+
+            if (actualType.IsAbstract)
+            {
+                string message = Error.Format(SRResources.CannotInstantiateAbstractResourceType, deltalink.TypeAnnotation.TypeName);
+                throw new ODataException(message);
+            }
+
+            IEdmEntityType actualEntityType = actualType as IEdmEntityType;
+
+            IEdmDeltaLinkBase edmDeltaLink;
+            ODataDeltaDeletedLink deletedLink = deltaLinkWrapper.DeltaLink as ODataDeltaDeletedLink;
+            if (deletedLink != null)
+            {
+                edmDeltaLink = new EdmDeltaDeletedLink(actualEntityType);
+            }
+            else
+            {
+                edmDeltaLink = new EdmDeltaLink(actualEntityType);
+            }
+
+            edmDeltaLink.Source = deltalink.Source;
+            edmDeltaLink.Target = deltalink.Target;
+            edmDeltaLink.Relationship = deltalink.Relationship;
+
+            return edmDeltaLink;
+        }
+
+        private static EdmDeltaDeletedEntityObject DeSerializeDeletedEntity(ODataDeserializerContext readContext, ODataDeletedResource deletedResource)
+        {
+            if (deletedResource == null)
+            {
+                throw new ODataException("Deleted resource not present");
+            }
+
+            IEdmModel model = readContext.Model;
+
+            if (model == null)
+            {
+                throw Error.Argument("readContext", SRResources.ModelMissingFromReadContext);
+            }
+
+            string typeName = deletedResource.TypeName;
+
+            IEdmStructuredType actualType = model.FindType(typeName) as IEdmStructuredType;
+            if (actualType == null)
+            {
+                throw new ODataException(Error.Format(SRResources.ResourceTypeNotInModel, typeName));
+            }
+
+            if (actualType.IsAbstract)
+            {
+                string message = Error.Format(SRResources.CannotInstantiateAbstractResourceType, typeName);
+                throw new ODataException(message);
+            }
+
+            IEdmEntityType actualEntityType = actualType as IEdmEntityType;
+
+            EdmDeltaDeletedEntityObject deletedEntity = new EdmDeltaDeletedEntityObject(actualEntityType);
+
+            deletedEntity.Id = deletedResource.Id.ToString();
+            deletedEntity.Reason = deletedResource.Reason.Value;
```

Comment:
Actually, in 4.01 a deletedEntity can have all of the properties of the non-deleted entity of that type, and may be identified by its key properties rather than the Id.  This may need some refactoring in order to use the type-specific deserializer to deserialize the properties, but as a deleted entity object.  Or, you'll have to make sure that the .Id can get properly populated based on key values in the payload.

Reply:
Actually, with [https://github.com/OData/odata.net/pull/1919](https://github.com/OData/odata.net/pull/1919) you should always be able to get the Id of the deleted resource.  Should probably add an assert (at least) that deletedResource.Id is not null, and make sure you have tests that serialize with 4.01 deleted resources that use keys rather than ids for the deleted resources.

---
In reply to: [503435218](https://github.com/OData/WebApi/pull/2275#discussion_r503435218) [](ancestors = 503435218)